spock-spring/src/main/java/org/spockframework/spring/mock/SpockMockPostprocessor.java

```diff
@@ -153,12 +155,37 @@ private String getBeanName(ConfigurableListableBeanFactory beanFactory,
     if (existingBeans.size() == 1) {
       return existingBeans.iterator().next();
     }
+    String primaryCandidate = determinePrimaryCandidate(registry, existingBeans, mockDefinition.getTypeToMock());
+    if (primaryCandidate != null) {
+      return primaryCandidate;
+    }
     throw new IllegalStateException(
       "Unable to register mock bean " + mockDefinition.getTypeToMock()
         + " expected a single matching bean to replace but found "
         + existingBeans);
   }
 
+  private String determinePrimaryCandidate(BeanDefinitionRegistry registry, Collection<String> candidateBeanNames,
+                                           ResolvableType type) {
+    String primaryBeanName = null;
+    for (String candidateBeanName : candidateBeanNames) {
+      BeanDefinition beanDefinition = registry.getBeanDefinition(candidateBeanName);
+      if (beanDefinition.isPrimary()) {
+        if (primaryBeanName != null) {
+          throw new NoUniqueBeanDefinitionException(type.resolve(), candidateBeanNames.size(),
+            "more than one 'primary' bean found among candidates: "
+              + Collections.singletonList(candidateBeanNames));
```

Comment:
Why do you wrap the collection in another list?
And this should also be tested.

Reply:
You're right - I'll cover this scenario.

As for the first question I'll rephrase it a bit - why do I add this method when it's already present in this class :)

Although in a very slightly different form:
```
  private String determinePrimaryCandidate(BeanDefinitionRegistry registry,
                                           String[] candidateBeanNames, ResolvableType type) {
    String primaryBeanName = null;
    for (String candidateBeanName : candidateBeanNames) {
      BeanDefinition beanDefinition = registry.getBeanDefinition(candidateBeanName);
      if (beanDefinition.isPrimary()) {
        if (primaryBeanName != null) {
          throw new NoUniqueBeanDefinitionException(type.resolve(),
            candidateBeanNames.length,
            "more than one 'primary' bean found among candidates: "
              + asList(candidateBeanNames));
        }
        primaryBeanName = candidateBeanName;
      }
    }
    return primaryBeanName;
  }
```

I'll reuse the existing method and add a test(s) to make the coverage better. Thank you for reporting this.

Reply:
@leonard84 I'll need an assistance with writing tests assertions for "failure" cases. I get different error responses depnding on whether I run tests using `:test` or `:testCglib`.

Reply:
Have a look at https://github.com/spockframework/spock/blob/master/spock-spring/src/test/groovy/org/spockframework/spring/mock/imported/SpringSpyWithAopProxyTests.groovy#L50-L55 there where other instances where ByteBuddy and Cglib behaved differently, maybe this helps. Could also check if those test cases need updating, they where all ported from spring boot.

Reply:
In my case Spring context is not loaded (which is expected) but when I use cglib, the actual failure of the nested test is `MultipleFailuresError` aggregating: context loading exception and `ClassPathException` ("org.spockframework.mock.CannotCreateMockException: Cannot create mock for class org.spockframework.spring.Service2").

The suggested `@PendingFeatureIf` usage would just make the test ignored which is not what I want.
My second idea was to check the system prop "ignoreByteBuddy" in the "then:" section and make assertions for expected failures for both cases but it turned out that cglib throws mentioned `ClassPathException` only when the feature is run within a bigger suite. When the the test was run alone, it behave the same way as with bytebuddy.

I though about writing a custom assertion method (just for this scenario) which would check if the failure was either the expected context initialization exception or `MultipleFailuresError` wrapping it but it turned out to be quite ugly.

I decided to just go with relaxed verification and just check if the error message contains "Failed to load ApplicationContext".

As for the test you've linked - it still doesn't pass for bytebuddy, where AoP (caching in that cased) gets messed up. It seems to have a different cause. I could try to port [the commit](https://github.com/spring-projects/spring-boot/commit/cdfbf28099b30621a118ebf2ec4988364295b610#diff-453b385bff6088b749d762f45d50b90aeed5799fb8d771ea614132997d0ca2a1), which closed [the related issue](https://github.com/spring-projects/spring-boot/issues/5837) but it extends the public API (new param for `SpringBean` annotation) and therefore deserves a separate PR.